ARCs/arc-0004.md

```diff
@@ -221,23 +243,40 @@ For example:
 A Contract is the complete set of methods that an app implements. It
 is similar to an interface, but may include further details about the
 concrete implementation. All methods in a Contract must be unique
-(specifically, they must have unique method selectors). Method names
-in Contracts **MAY** begin with underscore, but these names are
-reserved for use by this ARC and its future revisions.
-
-A Contract may contain two special methods, `_optIn` and
-`_closeOut`. These methods describe how callers can opt-in to
-(allocate local state), or close-out from (deallocate local state),
-such apps.  They may have any arguments or return values, but there
-must be no more than one method with each name.
-
-In the absence of such methods, ARC-4 Contracts **SHOULD** allow
-accounts to perform the corresponding action with no arguments, or by
-setting the associated OnCompletion value (OptIn = 1, CloseOut = 2)
-while calling a method of the contract. ARC-4 Contracts **MAY** allow
-OnCompletion values to be set to other values in order to support
-methods that delete or update the contract. Of course, great care
-should be taken when allowing these operations.
+(specifically, they must have unique method selectors).
+
+Methods in a Contract that originate from an implemented Interface **MUST**
+contain every allowed OnCompletion action from the Interface's definition
+of the method (though the Contract's program **MAY** reject some of these
+OnCompletion actions at runtime if they do not apply to the Contract's
+implementation). Additionally, the Contract's method definition **MAY**
+include additional allowed OnCompletion actions if its implementation
+supports them.
+
+Method names in Contracts **MAY** begin with underscore, but these
+names are reserved for use by this ARC and its future revisions.
+
+A Contract **MAY** implement a special method with the name
+`_create`. This method describes how a caller can instantiate a
```

Comment:
Is the `OnCompletion` restricted to `NoOp`?

Reply:
How can a contract support creation?  You would have to supply the bytecode itself.  So the contract would have to specify the bytecode. How else could you call it?

Reply:
Not just noop. Any `OnCompletion` value except close out or clear state can be specified for this, as those are the only ones that the network will reject.

Otherwise it's perfectly valid to opt-in during creation, or even set `OnCompletion` to delete to immediately self destruct the app. It's also valid to set `OnCompletion` to update, though this is functionally the same as noop.